### PR TITLE
[SPARK-46772][SQL][TESTS] Benchmarking Avro with Compression Codecs

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR improves AvroWriteBenchmark by adding benchmarks with codec and their extra functionalities.

- Avro compression with different codec
- Avro deflate/xz/zstandard with different levels
  - buffer pool if zstandard

### Why are the changes needed?

performance observation.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

connector/avro/src/test/scala/org/apache/spark/sql/execution/benchmark/AvroWriteBenchmark.scala

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

File: connector/avro/src/test/scala/org/apache/spark/sql/execution/benchmark/AvroWriteBenchmark.scala [hunk 1]
```diff
@@ -42,8 +44,8 @@ object AvroWriteBenchmark extends DataSourceWriteBenchmark {
     withTempPath { dir =>
       withTempTable("t1") {
         val width = 1000
-        val values = 500000
-        val files = 20
+        val values = 100000
```
Comment (yaooqinn) [inline connector/avro/src/test/scala/org/apache/spark/sql/execution/benchmark/AvroWriteBenchmark.scala:47]: Make it smaller as bzip2 took about 20 minutes to compress the original size of data

File: connector/avro/src/test/scala/org/apache/spark/sql/execution/benchmark/AvroWriteBenchmark.scala [hunk 2]
```diff
@@ -52,12 +54,48 @@ object AvroWriteBenchmark extends DataSourceWriteBenchmark {
...
+        bm.run()
+
+        AvroCompressionCodec.values().filter(_.getSupportCompressionLevel).foreach { codec =>
+          val bm = new Benchmark(
+            s"Avro ${codec.getCodecName} with different levels", values, output = output)
+          Seq(1, 3, 5, 7, 9).foreach { level =>
+            val conf = Map(s"spark.sql.avro.${codec.getCodecName}.level" -> level.toString)
+            addBenchmark(bm, codec.name, conf)
+            if (codec == AvroCompressionCodec.ZSTANDARD) {
+              val nondft =
```
Comment (dongjoon-hyun) [inline connector/avro/src/test/scala/org/apache/spark/sql/execution/benchmark/AvroWriteBenchmark.scala:89]: Does this mean `non-default`?

**Review by dongjoon-hyun — APPROVED** [review]
Thank you, @yaooqinn . Merged to master.

File: connector/avro/benchmarks/AvroWriteBenchmark-jdk21-results.txt [hunk 3]
```diff
@@ -1,16 +1,56 @@
...
+Output Single Int Column                           1433           1505         101         11.0          91.1       1.0X
+Output Single Double Column                        1467           1487          28         10.7          93.3       1.0X
+Output Int and String Column                       3187           3203          23          4.9         202.6       0.4X
+Output Partitions                                  2759           2796          52          5.7         175.4       0.5X
+Output Buckets                                     3760           3767           9          4.2         239.1       0.4X
 
-OpenJDK 64-Bit Server VM 21.0.1+12-LTS on Linux 5.15.0-1053-azure
+OpenJDK 64-Bit Server VM 21.0.2+13-LTS on Linux 5.15.0-1053-azure
 AMD EPYC 7763 64-Core Processor
-Write wide rows into 20 files:            Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply connector/avro/benchmarks/AvroWriteBenchmark-jdk21-results.txt:13]: Hi @dongjoon-hyun, I did not remove it as it now becomes the SNAPPY in the codec comparison benchmarking group

**Comment by yaooqinn** [comment]
Thank you @dongjoon-hyun

File: connector/avro/src/test/scala/org/apache/spark/sql/execution/benchmark/AvroWriteBenchmark.scala (hunk 2 above)
Comment (yaooqinn) [reply connector/avro/src/test/scala/org/apache/spark/sql/execution/benchmark/AvroWriteBenchmark.scala:89]: Yes, I hope the name doesn't bother you too much.：）

**Pull request closed without merging**